src/core/dataset.jl

```diff
@@ -0,0 +1,148 @@
+abstract type AbstractDataset end
+
+mutable struct DataFrameDataset <: AbstractDataset
+    values::DataFrames.DataFrame
+    timestamps::Vector{Dates.DateTime}
+    # Resolution is needed because AbstractDataset might have just one entry
+    resolution::Dates.Period
+    end_of_step_index::Int
+    last_recorded_row::Int
+    update_timestamp::Dates.DateTime
+end
+
+function DataFrameDataset(
+    values::DataFrames.DataFrame,
+    timestamps::Vector{Dates.DateTime},
+    resolution::Dates.Period,
+    end_of_step_index::Int,
+)
+    return DataFrameDataset(
+        values,
+        timestamps,
+        resolution,
+        end_of_step_index,
+        0,
+        UNSET_INI_TIME,
+    )
+end
+
+function DataFrameDataset(values::DataFrames.DataFrame)
+    return DataFrameDataset(
+        values,
+        Vector{Dates.DateTime}(),
+        Dates.Second(0.0),
+        1,
+        0,
+        UNSET_INI_TIME,
+    )
+end
+
+function make_system_state(
+    values::DataFrames.DataFrame,
+    timestamp::Dates.DateTime,
+    resolution::Dates.Period,
+)
+    return DataFrameDataset(values, [timestamp], resolution, 0, 1, UNSET_INI_TIME)
+end
+
+function get_dataset_value(s::DataFrameDataset, date::Dates.DateTime)
+    s_index = find_timestamp_index(get_timestamps(s), date)
+    if isnothing(s_index)
+        error("Request time stamp $date not in the state")
+    end
+    return get_values(s)[s_index, :]
+end
+
+function get_column_names(::OptimizationContainerKey, s::DataFrameDataset)
+    return DataFrames.names(get_values(s))
+end
+
+function get_last_recorded_value(s::DataFrameDataset)
+    if get_last_recorded_row(s) == 0
+        error("The Dataset hasn't been written yet")
+    end
+    return s.values[get_last_recorded_row(s), :]
+end
+
+get_data_resolution(s::DataFrameDataset) = s.resolution
+
+function get_end_of_step_timestamp(s::AbstractDataset)
+    return get_timestamps(s)[s.end_of_step_index]
+end
+
+get_timestamps(s::DataFrameDataset) = s.timestamps
```

Comment:
Since this can't be part of the external interface, I suggest removing the function and always use dot access.

Reply:
I am still figuring out if I will need the timestamps here or not.